wgpu-hal/src/gles/queue.rs

```diff
@@ -25,17 +25,112 @@ fn is_3d_target(target: super::BindTarget) -> bool {
     }
 }
 
+/// Helper macro to define repetative `set_[something]_enabled` functions in Queue impl
+macro_rules! gl_modes_set_enabled_fn {
+    ($fn_name:ident, $var_name:ident, $gl_var:ident) => {
+        unsafe fn $fn_name(&self, enabled: bool) {
+            self.state.lock().unwrap().$var_name = enabled;
+            if enabled {
+                self.shared.context.enable(glow::$gl_var);
+            } else {
+                self.shared.context.disable(glow::$gl_var);
+            }
+        }
+    };
+}
+
 impl super::Queue {
-    unsafe fn reset_state(&self) {
+    // Add helper functions for enabling and disabling certain GL features so that we can keep track
+    // of the current enabled/disabled state
+    unsafe fn set_current_program(&self, program: Option<glow::Program>) {
+        self.shared.context.use_program(program);
+        self.state.lock().unwrap().current_program = program;
+    }
+    gl_modes_set_enabled_fn!(
+        set_depth_test_enabled,
+        current_depth_test_enabled,
+        DEPTH_TEST
+    );
+    gl_modes_set_enabled_fn!(
+        set_stencil_test_enabled,
+        current_stencil_test_enabled,
+        STENCIL_TEST
+    );
+    gl_modes_set_enabled_fn!(
+        set_scissor_test_enabled,
+        current_scissor_test_enabled,
+        SCISSOR_TEST
+    );
+    gl_modes_set_enabled_fn!(set_blend_enabled, current_blend_enabled, BLEND);
+    gl_modes_set_enabled_fn!(set_cull_face_enabled, current_cull_face_enabled, CULL_FACE);
```

Comment:
If we're keeping all this state around anyway, we should be using it to deduplicate state changes.

Reply:
When trying to deduplicate the changes, interestingly things stop working. Like the depth or blending not getting set correctly. I think that means our tracked state is somehow getting out of sync with the actual GL state. :/

There's not a way to query the actual GL state is there? That would be better than us tracking it outselves probably.